promql/engine.go

```diff
@@ -1128,9 +1178,15 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 					}
 				}
 				maxt := ts - offset
+				if selVS.Timestamp > 0 {
+					maxt = timestamp.FromFloatSeconds(selVS.Timestamp) - offset
+				}
 				mint := maxt - selRange
 				// Evaluate the matrix selector for this series for this step.
-				points = ev.matrixIterSlice(it, mint, maxt, points)
+				if selVS.Timestamp <= 0 || (selVS.Timestamp > 0 && ts == ev.startTimestamp) {
+					// We only evaluate once for the constant selector.
```

Comment:
To keep the changes minimal. The upper level will make the evaluator an instant query when it's not a selector or subquery.

Reply:
But you already have this code already up a level, this appears to be doing things it doesn't need to.

Reply:
Yeah I see it is not required now

Reply:
I now remember why I had this thing (i.e. evaluating matrix only once at this level) - it is so that the engine will work just fine without the optimisation of the step invariant node. So I am bringing this back again because it would be good to be working correct even without the step invariant node optimisation.